### PR TITLE
[Feature] Support auto import modules from registry.

### PR DESCRIPTION
**Pull request by xiexinch**
## Motivation

The registry now supports auto-import modules from the given location.

register_all_modules before running is no longer needed. The modules will be lazy-imported during building.

- [x]  This PR can be merged after https://github.com/open-mmlab/mmengine/pull/643. The MMEngine version should be updated.

Ref: https://github.com/open-mmlab/mmdetection/pull/9143

### CODE REVIEW

**Review by MeowZheng — COMMENTED** [review]
please modify minimum version of mmengine v0.4.0

File: mmseg/registry/registry.py [hunk 1]
```diff
@@ -32,52 +32,90 @@
 from mmengine.registry import Registry
 
 # manage all kinds of runners like `EpochBasedRunner` and `IterBasedRunner`
-RUNNERS = Registry('runner', parent=MMENGINE_RUNNERS)
+RUNNERS = Registry(
+    'runner', parent=MMENGINE_RUNNERS, locations=['mmseg.engine.runner'])
```
Comment (MeowZheng) [inline mmseg/registry/registry.py:36]: there is no mmseg.engine.runner in mmseg?

File: demo/image_demo.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ def main():
         '--title', default='result', help='The image identifier.')
     args = parser.parse_args()
 
-    register_all_modules()
+    init_default_scope('mmseg')
```
Comment (MeowZheng) [inline demo/image_demo.py:27]: init_model have add `init_default_scope('mmseg')`,  so there is no need to add it

File: demo/inference_demo.ipynb [hunk 3]
```diff
@@ -24,9 +24,10 @@
     "import mmcv\n",
     "import matplotlib.pyplot as plt\n",
     "from mmengine.model.utils import revert_sync_batchnorm\n",
+    "from mmengine.registry import init_default_scope\n",
     "from mmseg.apis import init_model, inference_model, show_result_pyplot\n",
-    "from mmseg.utils import register_all_modules\n",
-    "register_all_modules()"
+    "\n",
+    "init_default_scope('mmseg')"
```
Comment (MeowZheng) [inline demo/inference_demo.ipynb:30]: same problem as above

File: docs/zh_cn/advanced_guides/datasets.md [hunk 4]
```diff
@@ -9,9 +9,10 @@
 实例化 Cityscapes 训练数据集:
 
 ```python
+from mmengine.registry import init_default_scope
 from mmseg.datasets import CityscapesDataset
-from mmseg.utils import register_all_modules
-register_all_modules()
+
+init_default_scope('mmseg')
```
Comment (MeowZheng) [inline docs/zh_cn/advanced_guides/datasets.md:15]: same problem as above

Comment (xiexinch) [reply docs/zh_cn/advanced_guides/datasets.md:15]: > same problem as above

Here are example codes for initializing a dataset, I think this line should not be removed.

File: docs/en/user_guides/3_inference.md [hunk 5]
```diff
@@ -30,15 +30,14 @@ Returns:
 Example:
 
 ```python
+from mmengine.registry import init_default_scope
 from mmseg.apis import init_model
-from mmseg.utils import register_all_modules
+
+init_default_scope('mmseg')
```
Comment (MeowZheng) [inline docs/en/user_guides/3_inference.md:36]: I think `init_default_scope` can be removed

**Review by MeowZheng — APPROVED** [review]

**Pull request merged**